independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Beans.java

```diff
@@ -650,6 +650,20 @@ static void validateInterceptorDecorator(BeanInfo bean, List<Throwable> errors,
 
     static void validateBean(BeanInfo bean, List<Throwable> errors, Consumer<BytecodeTransformer> bytecodeTransformerConsumer,
             Set<DotName> classesReceivingNoArgsCtor) {
+        // validate injection points, this includes producer method params, disposer method params and so on
+        List<InjectionPointInfo> injectionPoints = new ArrayList<>();
+        injectionPoints.addAll(bean.getAllInjectionPoints());
+        if (bean.getDisposer() != null) {
+            injectionPoints.addAll(bean.getDisposer().getAllInjectionPoints());
```

Comment:
When you query the bean for all injection points, it doesn't take into consideration those in disposer method, everything else is included. Is this intended @mkouba?

Reply:
Seems intentional to me, because injection points of the disposer method are _not_ injection points of the bean.

Class-based beans have these injection points:

- parameters of bean constructor
- `@Inject` fields
- parameters of initializer methods

Producer method beans have these injection points:

- parameters

And this is what you get from `BeanInfo.getAllInjectionPoints()`.

There are other injection points:

- parameters of disposer methods
- parameters of observer methods
- that's probably all?

You account for disposers here, but not for observers, which should probably be added? (It is legal to inject `Event` into an observer method, right?)

Reply:
I think so. Disposer injection points do not conceptually belong to a producer bean; i.e. a `@Dependent` bean injected into a disposer method param is destroyed when the invocation completes.

Reply:
Humm, there's a test for raw `Event` in observer method parameter. Either that test fails for some other reason (it doesn't verify the error message), or observer method parameter must be present in `BeanInfo.getAllInjectionPoints()`...

Reply:
Hmm, it seems like we don't validate observer injection points at all?

Reply:
Ah, no we do in the `ObserverInfo.init()` method.

Reply:
We should put the validation logic there. And we should probably do the same for beans, i.e. fail during injection point initialization...

Reply:
> or observer method parameter must be present in BeanInfo.getAllInjectionPoints()...

IIRC that was the case and the reason why I asked about disposer IP not being present. Either way I intentionally tested that case as well :)

> We should put the validation logic there. And we should probably do the same for beans, i.e. fail during injection point initialization...

Fair enough, I can look into that.